# Module Five Assignment: Cartpole Problem
Review the code in this notebook and in the score_logger.py file in the *scores* folder (directory). Once you have reviewed the code, return to this notebook and select **Cell** and then **Run All** from the menu bar to run this code. The code takes several minutes to run.

# Starter Code

In [ ]:
import random  
import gym  
import numpy as np  
from collections import deque  
from keras.models import Sequential  
from keras.layers import Dense  
from keras.optimizers import Adam  
  
  
from scores.score_logger import ScoreLogger  
  
ENV_NAME = "CartPole-v1"  
  
GAMMA = 0.95  
LEARNING_RATE = 0.001  
  
MEMORY_SIZE = 1000000  
BATCH_SIZE = 20  
  
EXPLORATION_MAX = 1.0  
EXPLORATION_MIN = 0.01  
EXPLORATION_DECAY = 0.995  
  
  
class DQNSolver:  
  
    def __init__(self, observation_space, action_space):  
        self.exploration_rate = EXPLORATION_MAX  
  
        self.action_space = action_space  
        self.memory = deque(maxlen=MEMORY_SIZE)  
  
        self.model = Sequential()  
        self.model.add(Dense(24, input_shape=(observation_space,), activation="relu"))  
        self.model.add(Dense(24, activation="relu"))  
        self.model.add(Dense(self.action_space, activation="linear"))  
        self.model.compile(loss="mse", optimizer=Adam(lr=LEARNING_RATE))  
  
    def remember(self, state, action, reward, next_state, done):  
        self.memory.append((state, action, reward, next_state, done))  
  
    def act(self, state):  
        if np.random.rand() < self.exploration_rate:  
            return random.randrange(self.action_space)  
        q_values = self.model.predict(state)  
        return np.argmax(q_values[0])  
  
    def experience_replay(self):  
        if len(self.memory) < BATCH_SIZE:  
            return  
        batch = random.sample(self.memory, BATCH_SIZE)  
        for state, action, reward, state_next, terminal in batch:  
            q_update = reward  
            if not terminal:  
                q_update = (reward + GAMMA * np.amax(self.model.predict(state_next)[0]))  
            q_values = self.model.predict(state)  
            q_values[0][action] = q_update  
            self.model.fit(state, q_values, verbose=0)  
        self.exploration_rate *= EXPLORATION_DECAY  
        self.exploration_rate = max(EXPLORATION_MIN, self.exploration_rate)  
  
  
def cartpole():  
    env = gym.make(ENV_NAME)  
    score_logger = ScoreLogger(ENV_NAME)  
    observation_space = env.observation_space.shape[0]  
    action_space = env.action_space.n  
    dqn_solver = DQNSolver(observation_space, action_space)  
    run = 0  
    while True:  
        run += 1  
        state = env.reset()  
        state = np.reshape(state, [1, observation_space])  
        step = 0  
        while True:  
            step += 1  
            #env.render()  
            action = dqn_solver.act(state)  
            state_next, reward, terminal, info = env.step(action)  
            reward = reward if not terminal else -reward  
            state_next = np.reshape(state_next, [1, observation_space])  
            dqn_solver.remember(state, action, reward, state_next, terminal)  
            state = state_next  
            if terminal:  
                print ("Run: " + str(run) + ", exploration: " + str(dqn_solver.exploration_rate) + ", score: " + str(step))  
                score_logger.add_score(step, run)  
                break  
            dqn_solver.experience_replay()  



In [2]:
cartpole()

Run: 1, exploration: 0.990025, score: 22
Scores: (min: 22, avg: 22, max: 22)

Run: 2, exploration: 0.918316468354365, score: 16
Scores: (min: 16, avg: 19, max: 22)

Run: 3, exploration: 0.7666961448653229, score: 37
Scores: (min: 16, avg: 25, max: 37)

Run: 4, exploration: 0.7183288830986236, score: 14
Scores: (min: 14, avg: 22.25, max: 37)

Run: 5, exploration: 0.6763948591909945, score: 13
Scores: (min: 13, avg: 20.4, max: 37)

Run: 6, exploration: 0.6337242817644086, score: 14
Scores: (min: 13, avg: 19.333333333333332, max: 37)

Run: 7, exploration: 0.5290920728090721, score: 37
Scores: (min: 13, avg: 21.857142857142858, max: 37)

Run: 8, exploration: 0.46912134373457726, score: 25
Scores: (min: 13, avg: 22.25, max: 37)

Run: 9, exploration: 0.4417353564707963, score: 13
Scores: (min: 13, avg: 21.22222222222222, max: 37)

Run: 10, exploration: 0.42013897252428334, score: 11
Scores: (min: 11, avg: 20.2, max: 37)

Run: 11, exploration: 0.3976004408064698, score: 12
Scores: (min: 11, a

Run: 89, exploration: 0.01, score: 21
Scores: (min: 9, avg: 176.8988764044944, max: 500)

Run: 90, exploration: 0.01, score: 172
Scores: (min: 9, avg: 176.84444444444443, max: 500)

Run: 91, exploration: 0.01, score: 119
Scores: (min: 9, avg: 176.2087912087912, max: 500)

Run: 92, exploration: 0.01, score: 229
Scores: (min: 9, avg: 176.7826086956522, max: 500)

Run: 93, exploration: 0.01, score: 194
Scores: (min: 9, avg: 176.96774193548387, max: 500)

Run: 94, exploration: 0.01, score: 124
Scores: (min: 9, avg: 176.40425531914894, max: 500)

Run: 95, exploration: 0.01, score: 104
Scores: (min: 9, avg: 175.6421052631579, max: 500)

Run: 96, exploration: 0.01, score: 167
Scores: (min: 9, avg: 175.55208333333334, max: 500)

Run: 97, exploration: 0.01, score: 158
Scores: (min: 9, avg: 175.37113402061857, max: 500)

Run: 98, exploration: 0.01, score: 150
Scores: (min: 9, avg: 175.1122448979592, max: 500)

Run: 99, exploration: 0.01, score: 332
Scores: (min: 9, avg: 176.6969696969697, max: 5

NameError: name 'exit' is not defined

Note: If the code is running properly, you should begin to see output appearing above this code block. It will take several minutes, so it is recommended that you let this code run in the background while completing other work. When the code has finished, it will print output saying, "Solved in _ runs, _ total runs."

You may see an error about not having an exit command. This error does not affect the program's functionality and results from the steps taken to convert the code from Python 2.x to Python 3. Please disregard this error.

# Experiment

Changes to parameters:

`GAMMA` - Decreased to 0.9.

`LEARNING_RATE` - Increased to 0.005.

`EXPLORATION_MIN` - Decreased to 0.005.

`EXPLORATION_DECAY` - Decreased to 0.98.

`BATCH_SIZE` - Increased to 32.

In [ ]:
import random  
import gym  
import numpy as np  
from collections import deque  
from keras.models import Sequential  
from keras.layers import Dense  
from keras.optimizers import Adam  
  
  
from scores.score_logger import ScoreLogger  
  
ENV_NAME = "CartPole-v1"  
  
GAMMA = 0.9
LEARNING_RATE = 0.005
  
MEMORY_SIZE = 1000000  
BATCH_SIZE = 32
  
EXPLORATION_MAX = 0.9
EXPLORATION_MIN = 0.005  
EXPLORATION_DECAY = 0.98
  
  
class DQNSolver:  
  
    def __init__(self, observation_space, action_space):  
        self.exploration_rate = EXPLORATION_MAX  
  
        self.action_space = action_space  
        self.memory = deque(maxlen=MEMORY_SIZE)  
  
        self.model = Sequential()  
        self.model.add(Dense(24, input_shape=(observation_space,), activation="relu"))  
        self.model.add(Dense(24, activation="relu"))  
        self.model.add(Dense(self.action_space, activation="linear"))  
        self.model.compile(loss="mse", optimizer=Adam(lr=LEARNING_RATE))  
  
    def remember(self, state, action, reward, next_state, done):  
        self.memory.append((state, action, reward, next_state, done))  
  
    def act(self, state):  
        if np.random.rand() < self.exploration_rate:  
            return random.randrange(self.action_space)  
        q_values = self.model.predict(state)  
        return np.argmax(q_values[0])  
  
    def experience_replay(self):  
        if len(self.memory) < BATCH_SIZE:  
            return  
        batch = random.sample(self.memory, BATCH_SIZE)  
        for state, action, reward, state_next, terminal in batch:  
            q_update = reward  
            if not terminal:  
                q_update = (reward + GAMMA * np.amax(self.model.predict(state_next)[0]))  
            q_values = self.model.predict(state)  
            q_values[0][action] = q_update  
            self.model.fit(state, q_values, verbose=0)  
        self.exploration_rate *= EXPLORATION_DECAY  
        self.exploration_rate = max(EXPLORATION_MIN, self.exploration_rate)  
  
  
def cartpole():  
    env = gym.make(ENV_NAME)  
    score_logger = ScoreLogger(ENV_NAME)  
    observation_space = env.observation_space.shape[0]  
    action_space = env.action_space.n  
    dqn_solver = DQNSolver(observation_space, action_space)  
    run = 0  
    while True:  
        run += 1  
        state = env.reset()  
        state = np.reshape(state, [1, observation_space])  
        step = 0  
        while True:  
            step += 1  
            #env.render()  
            action = dqn_solver.act(state)  
            state_next, reward, terminal, info = env.step(action)  
            reward = reward if not terminal else -reward  
            state_next = np.reshape(state_next, [1, observation_space])  
            dqn_solver.remember(state, action, reward, state_next, terminal)  
            state = state_next  
            if terminal:  
                print ("Run: " + str(run) + ", exploration: " + str(dqn_solver.exploration_rate) + ", score: " + str(step))  
                score_logger.add_score(step, run)  
                break  
            dqn_solver.experience_replay()  



In [ ]:
cartpole()

Run: 1, exploration: 0.9, score: 32
Scores: (min: 32, avg: 32, max: 32)

Run: 2, exploration: 0.6514179485332461, score: 17
Scores: (min: 17, avg: 24.5, max: 32)

Run: 3, exploration: 0.4348918153851448, score: 21
Scores: (min: 17, avg: 23.333333333333332, max: 32)

Run: 4, exploration: 0.35533827628916104, score: 11
Scores: (min: 11, avg: 20.25, max: 32)

Run: 5, exploration: 0.29033724280216744, score: 11
Scores: (min: 11, avg: 18.4, max: 32)

Run: 6, exploration: 0.25204917372901403, score: 8
Scores: (min: 8, avg: 16.666666666666668, max: 32)

Run: 7, exploration: 0.20594252585245243, score: 11
Scores: (min: 8, avg: 15.857142857142858, max: 32)

Run: 8, exploration: 0.1552069250156874, score: 15
Scores: (min: 8, avg: 15.75, max: 32)

Run: 9, exploration: 0.05885220348840902, score: 49
Scores: (min: 8, avg: 19.444444444444443, max: 49)

Run: 10, exploration: 0.021869579711016118, score: 50
Scores: (min: 8, avg: 22.5, max: 50)

Run: 11, exploration: 0.011690922547766706, score: 32
Sco

Run: 92, exploration: 0.005, score: 104
Scores: (min: 8, avg: 103.8586956521739, max: 500)

Run: 93, exploration: 0.005, score: 33
Scores: (min: 8, avg: 103.09677419354838, max: 500)

Run: 94, exploration: 0.005, score: 31
Scores: (min: 8, avg: 102.32978723404256, max: 500)

Run: 95, exploration: 0.005, score: 133
Scores: (min: 8, avg: 102.65263157894736, max: 500)

Run: 96, exploration: 0.005, score: 46
Scores: (min: 8, avg: 102.0625, max: 500)

Run: 97, exploration: 0.005, score: 50
Scores: (min: 8, avg: 101.52577319587628, max: 500)

Run: 98, exploration: 0.005, score: 175
Scores: (min: 8, avg: 102.27551020408163, max: 500)

Run: 99, exploration: 0.005, score: 89
Scores: (min: 8, avg: 102.14141414141415, max: 500)

Run: 100, exploration: 0.005, score: 75
Scores: (min: 8, avg: 101.87, max: 500)

Run: 101, exploration: 0.005, score: 11
Scores: (min: 8, avg: 101.66, max: 500)

Run: 102, exploration: 0.005, score: 11
Scores: (min: 8, avg: 101.6, max: 500)

Run: 103, exploration: 0.005, 

Run: 194, exploration: 0.005, score: 28
Scores: (min: 8, avg: 19.84, max: 175)

Run: 195, exploration: 0.005, score: 16
Scores: (min: 8, avg: 18.67, max: 175)

Run: 196, exploration: 0.005, score: 19
Scores: (min: 8, avg: 18.4, max: 175)

Run: 197, exploration: 0.005, score: 20
Scores: (min: 8, avg: 18.1, max: 175)

Run: 198, exploration: 0.005, score: 14
Scores: (min: 8, avg: 16.49, max: 170)

Run: 199, exploration: 0.005, score: 12
Scores: (min: 8, avg: 15.72, max: 170)

Run: 200, exploration: 0.005, score: 32
Scores: (min: 8, avg: 15.29, max: 170)

Run: 201, exploration: 0.005, score: 41
Scores: (min: 8, avg: 15.59, max: 170)

Run: 202, exploration: 0.005, score: 14
Scores: (min: 8, avg: 15.62, max: 170)

Run: 203, exploration: 0.005, score: 18
Scores: (min: 8, avg: 15.71, max: 170)

Run: 204, exploration: 0.005, score: 12
Scores: (min: 8, avg: 15.72, max: 170)

Run: 205, exploration: 0.005, score: 9
Scores: (min: 8, avg: 15.54, max: 170)

Run: 206, exploration: 0.005, score: 11
Sco

KeyboardInterrupt: 

The goal of the agent within the cartpole problem is to keep a pole balanced upright in a cart for as long as possible. If the pole falls past a certain angle or the cart goes beyond the bounds of the track, the agent loses. The different states are the cart position, cart velocity, pole angle, and pole angular velocity. These values are represented within the program as a state vector. An agent within this simulation has two possible actions, push the cart to left or push the cart to the right. The reinforment learning algorithm being used is Deep Q-Learning (DQN). DQN is an improved version of Q-Learning. It enables agents to learn smarter and faster by leveraging a neural network instead of a large lookup table, balancing exploration and exploitation, remembering and recalling past experience, and using networks to avoid wild updates (Beysolow, 2019).

The purpose of experience replay is to improve training by letting the agent learn from past experience. In this algorithm, a replay buffer of size one million is leveraged to store experiences. Every time a step completes, the agent uses the `remember()` method to save to this memory. At a certain point, the `experience_replay()` method is called to recall a random batch of experiences. For this random batch of experiences, the Q-value is updated using the Bellman equation. The model is then trained to better it's Q-value prediction. A discount factor (`GAMMA`) is also introduced which determines how much the agent will favor short-term vs. long-term rewards. In the starter code, this is set to `0.95` which encourages the agent to choose sustained rewards over time while still accounting for short-term outcomes.

The artificial neural network used in this algorithm has an input layer, two hidden layers, and an output layer. It uses the Adam optimizer and the MSE loss function. The input layer takes in a state vector of cart position, cart velocity, pole angle, and pole angular velocity. The first and seccond hidden layers have twenty-four neurons and use the ReLU activation function. The output layer has two neurons, one for move left and one for move right, and uses a linear activation function. It outputs a Q-value which represents the expected future reward of each action at the current state. By leveraging a neural network, the algorithm becomes more efficient. Since the algorithm can approximate Q-values without using a lookup table, it does not have to memorize each state-action pair in a massive lookup table which would prove nigh impossible in environments with a large number of possible states and actions (Singh, 2022).

When increasing the learning rate, along with several other changes to the starting parameters, the algorithm degraded significanly resulting in failure to achieve the start code's learning rate of `>=` 195 average score after 109 runs. The average score hovered at around ~100 from runs 50 to 120. After that, the average score plummeted to ~20. I eventually killed the program after run 239. By decreasing the `EXPLORATION_DECAY`, the agent switched from exploration to exploitation even though it was likely undertrained and stuck in a subopmtimal policy. I also increased the `LEARNING_RATE`, which further amplified the issue. I also lowered the `GAMMA` which caused the agent to favor short-term rewards. Finally, the `BATCH_SIZE` was increased which further amplified the instability. In summary, the agent stopped exploring too soon and learned a bad policy too aggressively.

**Resources**

Beysolow, T. II. (2019). *Applied reinforcement learning with Python: With OpenAI Gym, TensorFlow, and Keras* (1st ed.). Apress.

Singh, S. (2022, October 7) *How are neural networks used in deep Q-learning?* Turing. Retrieved July 31, 2025, from https://www.turing.com/kb/how-are-neural-networks-used-in-deep-q-learning